pkg/oneagent/oneagent.go

```diff
@@ -0,0 +1,237 @@
+		// do not return err as it would trigger yet another reconciliation loop immediately
+		return nil
+	} else if desired != "" && oneagent.Status.Version != desired {
+		logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "previous": oneagent.Status.Version, "desired": desired}).Info("new version available")
+		oneagent.Status.Version = desired
+		updateStatus = true
+	}
+
+	// query oneagent pods
+	podList := util.BuildPodList()
+	labelSelector := labels.SelectorFromSet(oneagent.BuildLabels()).String()
+	listOps := &metav1.ListOptions{LabelSelector: labelSelector}
+	err = query.List(oneagent.Namespace, podList, query.WithListOptions(listOps))
+	if err != nil {
+		logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "pods": podList, "error": err}).Error("failed to query pods")
+		return err
+	}
+
+	// determine pods to restart
+	podsToDelete, instances := util.GetPodsToRestart(podList.Items, dtc, oneagent)
+	if !reflect.DeepEqual(instances, oneagent.Status.Items) {
+		logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "status.items": instances}).Info("status changed")
+		updateStatus = true
+		oneagent.Status.Items = instances
+	}
+
+	// restart daemonset
+	err = deletePods(oneagent, podsToDelete)
+	if err != nil {
+		logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "error": err}).Error("failed to delete pods")
+		return err
+	}
+
+	// update status
+	if updateStatus {
+		oneagent.Status.UpdatedTimestamp = metav1.Now()
+		logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "status": oneagent.Status}).Info("updating status")
+		err := action.Update(oneagent)
+		if err != nil {
+			logrus.WithFields(logrus.Fields{"oneagent": oneagent.Name, "error": err}).Error("failed to update status")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deletePods deletes a list of pods
+//
+// Returns an error in the following conditions:
+//  - failure on object deletion
+//  - timeout on waiting for ready state
+func deletePods(cr *v1alpha1.OneAgent, pods []corev1.Pod) error {
+	for _, pod := range pods {
+		// delete pod
+		logrus.WithFields(logrus.Fields{"oneagent": cr.Name, "pod": pod.Name, "nodeName": pod.Spec.NodeName}).Info("deleting pod")
+		err := action.Delete(&pod)
+		if err != nil {
+			logrus.WithFields(logrus.Fields{"oneagent": cr.Name, "pod": pod.Name, "error": err}).Error("failed to delete pod")
+			return err
+		}
+
+		// wait for pod on node to get "Running" again
+		var status error
+		fieldSelector, _ := fields.ParseSelector(fmt.Sprintf("spec.nodeName=%v,status.phase=Running,metadata.name!=%v", pod.Spec.NodeName, pod.Name))
+		labelSelector := labels.SelectorFromSet(cr.BuildLabels())
+		logrus.WithFields(logrus.Fields{"field-selector": fieldSelector, "label-selector": labelSelector}).Debug("query pod")
+		listOps := &metav1.ListOptions{FieldSelector: fieldSelector.String(), LabelSelector: labelSelector.String()}
+		for splay := uint16(0); splay < *cr.Spec.WaitReadySeconds; splay += splayTimeSeconds {
```

Comment:
`for` within `for` makes it hard follow, rethink if we can refactor this handling to a function to get it less complex.